pytensor/link/jax/ops.py

```diff
@@ -0,0 +1,472 @@
+    >>> print(f(1, 2))
+    [array(3.)]
+
+    Notes
+    -----
+    The function is based on a blog post by Ricardo Vieira and Adrian Seyboldt,
+    available at
+    `pymc-labls.io <https://www.pymc-labs.io/blog-posts/jax-functions-in-pymc-3-quick
+    -examples/>`__.
+    To accept functions and non pytensor variables as input, the function make use
+    of :func:`equinox.partition` and :func:`equinox.combine` to split and combine the
+    variables. Shapes are inferred using
+    :func:`pytensor.compile.builders.infer_shape` and :func:`jax.eval_shape`.
+
+    """
+
+    def func(*args, **kwargs):
+        # 1. Partition inputs into dynamic pytensor variables, wrapped functions and
+        # static variables.
+        # Static variables don't take part in the graph.
+        pt_vars, func_vars, static_vars = _split_inputs(args, kwargs)
+
+        # 2. Get the original variables from the wrapped functions.
+        vars_from_func = tree_map(lambda f: f.get_vars(), func_vars)
+        input_dict = {"vars": pt_vars, "vars_from_func": vars_from_func}
+
+        # 3. Flatten the input dictionary.
+        # e.g. {"a": tensor_a, "b": [tensor_b]} becomes [tensor_a, tensor_b], because
+        # pytensor ops only accepts lists of pytensor.Variables as input.
+        pt_vars_flat, pt_vars_treedef = tree_flatten(
+            input_dict,
+        )
+        pt_types = [var.type for var in pt_vars_flat]
+
+        # 4. Create dummy inputs for shape inference.
+        shapes = _infer_shapes(pt_vars_flat, use_infer_static_shape)
+        dummy_in_flat = _create_dummy_inputs_from_shapes(
+            pt_vars_flat, shapes, use_infer_static_shape
+        )
+        dummy_inputs = tree_unflatten(pt_vars_treedef, dummy_in_flat)
+
+        # 5. Partition the JAX function into dynamic and static parts.
+        jaxfunc_dynamic, static_out_dic = _partition_jaxfunc(
+            jaxfunc, static_vars, func_vars
+        )
+        flat_func = _flatten_func(jaxfunc_dynamic, pt_vars_treedef)
+
+        # 6. Infer output types using JAX's eval_shape.
+        out_treedef, pt_types_flat = _infer_output_types(jaxfunc_dynamic, dummy_inputs)
+
+        # 7. Create the Pytensor Op instance.
+        curr_name = "JAXOp_" + (jaxfunc.__name__ if name is None else name)
+        op_instance = JAXOp(
+            pt_types,
+            pt_types_flat,
+            flat_func,
+            name=curr_name,
+        )
+
+        # 8. Execute the op and unflatten the outputs.
+        output_flat = op_instance(*pt_vars_flat)
+        if not isinstance(output_flat, Sequence):
+            output_flat = [output_flat]
+        outvars = tree_unflatten(out_treedef, output_flat)
+
+        # 9. Combine with static outputs and wrap eventual output functions with
+        # _WrappedFunc
+        return _process_outputs(static_out_dic, jaxfunc, args, kwargs, outvars)
+
+    return func
+
+
+def _filter_ptvars(x):
+    return isinstance(x, pt.Variable)
```

Comment:
Doesn't this mean that if the user passes a numpy array it will be considered static? It's common for users to pass numpy objects to PyTensor operations. Should the constant input separation be made explicitly perhaps? Or even not part of the API, the user can pass a partialled JAX function to PyTensor (or inline the constants themselves)?